File: progra3-0583-equipo5-frontend/Frontend_C#/SoftPac/SoftPacBusiness/UsuariosBO.cs
using SoftPac.Model;
using SoftPac.Persistance.DAO;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SoftPac.Business
{
    public class UsuariosBO
    {
        private UsuariosDAO usuariosDAO;
        private UsuarioPaisAccesoDAO usuarioPaisAccesoDAO;

        public UsuariosBO()
        {
            this.usuariosDAO = new UsuariosDAOImpl();
            this.usuarioPaisAccesoDAO = new UsuarioPaisAccesoDAOImpl();
        }

        // --- MÉTODO NUEVO PARA AUTENTICAR ---
        public UsuariosDTO AutenticarUsuario(string nombreUsuario, string password)
        {
            // 1. Buscamos al usuario por su nombre de usuario
            UsuariosDTO usuario = this.usuariosDAO.ObtenerPorNombreUsuario(nombreUsuario);

            if (usuario == null)
            {
                return null; // Usuario no encontrado
            }

            // 2. Verificamos la contraseña
            // !! ADVERTENCIA DE SEGURIDAD !!
            // Esto es solo para la demostración con datos hardcodeados.
            // En un sistema real, NUNCA guardes contraseñas en texto plano.
            // Deberías usar una librería como BCrypt.Net para comparar el hash.
            // Ejemplo real: bool esValido = BCrypt.Net.BCrypt.Verify(password, usuario.PasswordHash);

            bool esPasswordValido = (password == usuario.PasswordHash);

            if (esPasswordValido)
            {
                return usuario; // Autenticación exitosa
            }

            return null; // Contraseña incorrecta
        }

        // Aquí puedes añadir el resto de los métodos que necesites (Insertar, Modificar, etc.)
        public IList<UsuariosDTO> ListarTodos()
        {
            return usuariosDAO.ListarTodos();
        }

        public UsuariosDTO ObtenerPorId(int usuarioId)
        {
            return usuariosDAO.ObtenerPorId(usuarioId);
        }

        public int InsertarUsuario(UsuariosDTO nuevoUsuario)
        {
            if (string.IsNullOrWhiteSpace(nuevoUsuario.PasswordHash))
            {
                throw new Exception("La contraseña es obligatoria para nuevos usuarios.");
            }
            // Simulación de Hashing. En un proyecto real, usar BCrypt.Net.HashPassword(nuevoUsuario.PasswordHash);
            nuevoUsuario.PasswordHash = nuevoUsuario.PasswordHash;

            return usuariosDAO.Insertar(nuevoUsuario);
        }

        public int ModificarAccesoUsuario(int usuarioId, string nuevoNombreUsuario, bool activo, List<int> paisesIds)
        {
            UsuariosDTO usuario = usuariosDAO.ObtenerPorId(usuarioId);
            if (usuario == null) return 0;

            usuario.NombreDeUsuario = nuevoNombreUsuario;
            usuario.Activo = activo;

            // Actualizar la lista de países de acceso
            usuario.UsuarioPais.Clear();
            foreach (var paisId in paisesIds)
            {
                usuario.UsuarioPais.Add(new UsuarioPaisAccesoDTO
                {
                    Usuario = new UsuariosDTO { UsuarioId = usuarioId },
                    Pais = new PaisesDTO { PaisId = paisId },
                    Acceso = true
                });
            }

            return usuariosDAO.Modificar(usuario);
        }

        public int EliminarUsuario(int usuarioId, int usuarioEliminacionId)
        {
            UsuariosDTO usuario = usuariosDAO.ObtenerPorId(usuarioId);
            if (usuario == null) return 0;

            UsuariosDTO usuarioEliminador = new UsuariosDTO { UsuarioId = usuarioEliminacionId };
            usuario.UsuarioEliminacion = usuarioEliminador;

            return usuariosDAO.eliminarLogico(usuario);
        }



    }
}

File: Backend_Java/SoftPacBusiness/src/main/java/pe/edu/pucp/softpac/bo/UsuariosBO.java
package pe.edu.pucp.softpac.bo;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.logging.Level;
import java.util.logging.Logger;
import pe.edu.pucp.softpac.bo.util.PasswordUtil;
import pe.edu.pucp.softpac.dao.UsuarioPaisAccesoDAO;
import pe.edu.pucp.softpac.dao.UsuariosDAO;
import pe.edu.pucp.softpac.daoImpl.UsuarioPaisAccesoDAOImpl;
import pe.edu.pucp.softpac.daoImpl.UsuariosDAOImpl;
import pe.edu.pucp.softpac.model.PaisesDTO;
import pe.edu.pucp.softpac.model.UsuarioPaisAccesoDTO;
import pe.edu.pucp.softpac.model.UsuariosDTO;
import pe.edu.pucp.softpac.bo.util.EmailUtil;
import pe.edu.pucp.softpac.bo.util.TipoOperacionUsuario;
import pe.edu.pucp.softpac.dao.PaisesDAO;
import pe.edu.pucp.softpac.daoImpl.PaisesDAOImpl;

public class UsuariosBO {

    private UsuariosDAO usuariosDAO;
    private UsuarioPaisAccesoDAO usuariosPaisAccesoDAO;

    public UsuariosBO() {
        this.usuariosDAO = new UsuariosDAOImpl();
        this.usuariosPaisAccesoDAO = new UsuarioPaisAccesoDAOImpl();
    }

    public UsuariosDTO autenticarUsuarioPorNombreUsuario(String nombreUsuario, String password) {
        UsuariosDTO usuario = this.usuariosDAO.obtenerPorNombreDeUsuario(nombreUsuario.trim());

        if (usuario == null) {
            return new UsuariosDTO(); //Usuario no encontrado
        }
        
        Boolean esValido = PasswordUtil.checkPassword(password.trim(), usuario.getPassword_hash());

        if (esValido) {
            //PARA SERIALIZACIÓN:
            for (UsuarioPaisAccesoDTO upa : usuario.getUsuario_pais()) {
                upa.setUsuario(null);
            }
            return usuario; //Autenticación exitosa
        }

        return new UsuariosDTO(); //Contraseña incorrecta
    }

    public UsuariosDTO autenticarUsuarioPorCorreo(String correo_usuario, String password) {
        UsuariosDTO usuario = this.usuariosDAO.obtenerPorCorreo(correo_usuario);

        if (usuario == null) {
            return null; //Usuario no encontrado
        }

        Boolean esValido = PasswordUtil.checkPassword(password.trim(), usuario.getPassword_hash());

        if (esValido) {
            //PARA SERIALIZACIÓN:
            for (UsuarioPaisAccesoDTO upa : usuario.getUsuario_pais()) {
                upa.setUsuario(null);
            }
            return usuario; //Autenticación exitosa
        }

        return null; //Contraseña incorrecta

    }

    public ArrayList<UsuariosDTO> listarTodos() {
        return (ArrayList<UsuariosDTO>) usuariosDAO.listarTodos();
    }


    public UsuariosDTO obtenerPorId(Integer usuario_id) {
        UsuariosDTO usuario = new UsuariosDTO();
        usuario.setUsuario_id(usuario_id);

        usuario = usuariosDAO.obtenerPorId(usuario_id);

        //PARA SERIALIZACIÓN:
        for (UsuarioPaisAccesoDTO upa : usuario.getUsuario_pais()) {
            upa.setUsuario(null);
        }

        return usuario;
    }


    public Integer insertarUsuario(UsuariosDTO nuevoUsuario, UsuariosDTO usuarioActual) {
        //1RO INSERTAR USUARIO
        if (nuevoUsuario.getPassword_hash() == null) {
            try {
                throw new Exception("La contraseña es obligatoria para nuevos usuarios.");
            } catch (Exception ex) {
                Logger.getLogger(UsuariosBO.class.getName()).log(Level.SEVERE, null, ex);
            }
        }
        String guardaContrasenhaOriginal = nuevoUsuario.getPassword_hash();
        String hashedPassword = PasswordUtil.hashPassword(nuevoUsuario.getPassword_hash());
        nuevoUsuario.setPassword_hash(hashedPassword);
        PaisesDAO paisesDAO = new PaisesDAOImpl();
        List<PaisesDTO> paises = paisesDAO.listarTodos();
        for (UsuarioPaisAccesoDTO acceso : nuevoUsuario.getUsuario_pais()) {
            acceso.getPais().setNombre(obtenerNombrePais(acceso.getPais().getPais_id(), paises));
            acceso.setUsuario(nuevoUsuario);
        }
        Integer resultado = usuariosDAO.insertar(nuevoUsuario);
        //2DO: CORREO
        ArrayList<UsuarioPaisAccesoDTO> accesosConcedidos = new ArrayList<>();
        for(UsuarioPaisAccesoDTO acceso : nuevoUsuario.getUsuario_pais()){
            if(acceso.getAcceso()==Boolean.TRUE) accesosConcedidos.add(acceso);
        }
        nuevoUsuario.setUsuario_pais(accesosConcedidos);
        if(resultado>0){
            EmailUtil eu = new EmailUtil();
            nuevoUsuario.setPassword_hash(guardaContrasenhaOriginal);
            eu.enviarCorreo(nuevoUsuario, usuarioActual, TipoOperacionUsuario.INSERTAR);
        }
        return resultado;

    }

    public Integer modificarAccesoUsuario(int usuarioId, String nuevoNombreUsuario,
            Boolean activo, List<Integer> paisesIds, UsuariosDTO usuarioActual,
            String nuevaContrasenha) {
        UsuariosDTO usuario = usuariosDAO.obtenerPorId(usuarioId);
        if (usuario == null) {
            return 0;
        }
        
        // --- GUARDAR ESTADO ANTERIOR DE ACCESOS ---
        Map<Integer, Boolean> accesosAnteriores = new HashMap<>();
        if (usuario.getUsuario_pais() != null) {
            for (UsuarioPaisAccesoDTO upa : usuario.getUsuario_pais()) {
                if (upa.getPais() != null) {
                    accesosAnteriores.put(
                        upa.getPais().getPais_id(), 
                        upa.getAcceso() != null ? upa.getAcceso() : false
                    );
                }
            }
        }
        
        //           ACTUALIZACION
        usuario.setNombre_de_usuario(nuevoNombreUsuario);
        usuario.setActivo(activo);
        
        //Catalogo de Paises para los nombres
        PaisesDAO paisesDAO = new PaisesDAOImpl();
        List<PaisesDTO> catalogoPaises = paisesDAO.listarTodos();
        
        //Actualizamos la lista de países
        usuario.getUsuario_pais().clear();
        for (PaisesDTO pais : catalogoPaises) {
            UsuarioPaisAccesoDTO user = new UsuarioPaisAccesoDTO();
            String nombrePais = this.obtenerNombrePais(pais.getPais_id(), catalogoPaises);
            pais.setNombre(nombrePais);
            user.setUsuario(usuario);
            user.setPais(pais);
            user.setAcceso(paisesIds.contains(pais.getPais_id()));
            usuario.getUsuario_pais().add(user);
        }
        String guardaContrasenhaOriginal = null;
        //Actualizamos contraseña
        if(nuevaContrasenha!=null && !nuevaContrasenha.trim().isEmpty()){
            guardaContrasenhaOriginal = nuevaContrasenha;
            String hashedPassword = PasswordUtil.hashPassword(nuevaContrasenha);
            usuario.setPassword_hash(hashedPassword);
        }
                
        //        FIN ACTUALIZACION
        
        Integer resultado = usuariosDAO.modificar(usuario);
        
        if(resultado > 0) {
            // --- FILTRAR SOLO LOS CAMBIOS DE ACCESO ---
            List<UsuarioPaisAccesoDTO> soloAccesosCambiados = new ArrayList<>();
            
            for (UsuarioPaisAccesoDTO upa : usuario.getUsuario_pais()) {
                Integer paisId = upa.getPais().getPais_id();
                Boolean accesoAnterior = accesosAnteriores.getOrDefault(paisId, false);
                Boolean accesoNuevo = upa.getAcceso() != null ? upa.getAcceso() : false;
                
                // Solo incluir si hubo un CAMBIO en el acceso
                if (!accesoAnterior.equals(accesoNuevo)) {
                    soloAccesosCambiados.add(upa);
                }
            }
            
            // Reemplazar temporalmente la lista completa con solo los cambios
            usuario.setUsuario_pais((ArrayList<UsuarioPaisAccesoDTO>) soloAccesosCambiados);
            usuario.setPassword_hash(guardaContrasenhaOriginal);
            
            EmailUtil eu = new EmailUtil();
            eu.enviarCorreo(usuario, usuarioActual, TipoOperacionUsuario.MODIFICARACCESO);
            
        }
        return resultado;
    }


    public Integer modificarUsuario(UsuariosDTO usuario) {
        UsuariosDTO usuarioExistente = usuariosDAO.obtenerPorId(usuario.getUsuario_id());
        if (usuarioExistente == null) {
            return 0;
        }

        // Actualizar campos básicos
        usuarioExistente.setNombre_de_usuario(usuario.getNombre_de_usuario());
        usuarioExistente.setActivo(usuario.getActivo());

        // Actualizar accesos a países
        usuarioExistente.getUsuario_pais().clear();

        for (UsuarioPaisAccesoDTO acceso : usuario.getUsuario_pais()) {
            acceso.setUsuario(usuarioExistente);
            usuarioExistente.getUsuario_pais().add(acceso);
        }

        return usuariosDAO.modificar(usuarioExistente);
    }

    public Integer eliminarUsuario(UsuariosDTO usuario, UsuariosDTO usuarioActual) {

        usuario.setFecha_eliminacion(new Date());
        usuario.setUsuario_eliminacion(usuarioActual);
        for (UsuarioPaisAccesoDTO acceso : usuario.getUsuario_pais()) {
            acceso.setUsuario(usuario);
        }
        Integer resultado = usuariosDAO.eliminarLogico(usuario);
        if(resultado>0){
            EmailUtil eu = new EmailUtil();
            eu.enviarCorreo(usuario, usuarioActual , TipoOperacionUsuario.ELIMINAR);
        }
        return resultado;

    }
    
    private String obtenerNombrePais(Integer paisId, List<PaisesDTO> paises){
        String nombrePais="Pais Desconocido";
        for(PaisesDTO pais : paises){
            if(Objects.equals(pais.getPais_id(), paisId)){
                nombrePais=pais.getNombre();
            }
        }
        return nombrePais;
    }

}